src/python_testing/TC_IDM_2_2.py

```diff
@@ -0,0 +1,733 @@
+
+from mobly import asserts
+
+import matter.clusters as Clusters
+from matter import ChipDeviceCtrl
+from matter.clusters import ClusterObjects as ClusterObjects
+from matter.clusters.Attribute import AttributePath
+from matter.exceptions import ChipStackError
+from matter.interaction_model import InteractionModelError, Status
+from matter.testing import global_attribute_ids
+from matter.testing.basic_composition import BasicCompositionTests
+from matter.testing.matter_testing import MatterBaseTest, TestStep, async_test_body, default_matter_test_main
+
+
+class TC_IDM_2_2(MatterBaseTest, BasicCompositionTests):
+    """Test case for IDM-2.2: Report Data Action from DUT to TH.
+
+    This test verifies that the DUT correctly handles read requests and responds
+    with appropriate report data actions. It covers various scenarios including:
+    - Reading single and multiple attributes
+    - Reading global and non-global attributes
+    - Handling unsupported endpoints/clusters/attributes
+    - Data version filtering
+    - Chunked data messages
+    - Access control
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.endpoint = 0
+
+    # === Attribute Reading and Verification ===
+    async def read_attribute(self, attribute_path: list) -> dict:
+        """Read attributes from the device."""
+        try:
+            return await self.default_controller.Read(
+                self.dut_node_id,
+                attribute_path)
+        except ChipStackError as e:  # chipstack-ok
+            if "Timeout" in str(e):
+                logging.error("Read operation timed out")
+                logging.error(f"Attribute path: {attribute_path}")
+            raise
+
+    async def verify_attribute_read(self, attribute_path: list, expected_error: Status = None) -> dict:
+        """Read and verify attributes from the device.
+
+        Args:
+            attribute_path: List of attribute paths to read (should be AttributePath objects or tuples like (endpoint, cluster, attribute))
+            expected_error: Optional expected error status
+
+        Returns:
+            Dictionary containing the read results
+
+        Raises:
+            AssertionError if verification fails
+        """
+        if expected_error:
+            try:
+                if isinstance(attribute_path[0], AttributePath):
+                    endpoint = attribute_path[0].EndpointId
+                    cluster = attribute_path[0].ClusterId
+                    attribute = attribute_path[0].AttributeId
+                else:
+                    # Handle tuple case
+                    endpoint = attribute_path[0][0] if len(attribute_path[0]) > 0 else None
+                    cluster = attribute_path[0][1] if len(attribute_path[0]) > 1 else None
+                    attribute = attribute_path[0][2] if len(attribute_path[0]) > 2 else None
+                result = await self.read_single_attribute_expect_error(

```

Comment:
You'll get a KeyError here if you pass in anything other than a single attribute. I don't think this code is working as expected.

read_single_attribute_expect_error uses the endpoints as keys to read the returned attribute paths. This will throw a key error if the key is None. That doesn't mean the ReadAttribute call failed with the expected error.